File: rest/src/main/java/io/sparkled/rest/service/sequence/SequenceRestServiceHandler.java
package io.sparkled.rest.service.sequence;

import com.google.inject.persist.Transactional;
import io.sparkled.model.entity.*;
import io.sparkled.model.render.RenderedStagePropDataMap;
import io.sparkled.model.util.SequenceUtils;
import io.sparkled.model.validator.exception.EntityNotFoundException;
import io.sparkled.persistence.sequence.SequencePersistenceService;
import io.sparkled.persistence.song.SongPersistenceService;
import io.sparkled.persistence.stage.StagePersistenceService;
import io.sparkled.renderer.Renderer;
import io.sparkled.rest.response.IdResponse;
import io.sparkled.rest.service.RestServiceHandler;
import io.sparkled.viewmodel.sequence.SequenceViewModel;
import io.sparkled.viewmodel.sequence.SequenceViewModelConverter;
import io.sparkled.viewmodel.sequence.channel.SequenceChannelViewModel;
import io.sparkled.viewmodel.sequence.channel.SequenceChannelViewModelConverter;
import io.sparkled.viewmodel.sequence.search.SequenceSearchViewModel;
import io.sparkled.viewmodel.sequence.search.SequenceSearchViewModelConverter;
import io.sparkled.viewmodel.stage.StageViewModel;
import io.sparkled.viewmodel.stage.StageViewModelConverter;
import io.sparkled.viewmodel.stage.prop.StagePropViewModel;
import io.sparkled.viewmodel.stage.prop.StagePropViewModelConverter;

import javax.inject.Inject;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.toList;

class SequenceRestServiceHandler extends RestServiceHandler {

    private static final String MP3_MIME_TYPE = "audio/mpeg";

    private final SequencePersistenceService sequencePersistenceService;
    private final SongPersistenceService songPersistenceService;
    private final StagePersistenceService stagePersistenceService;
    private final SequenceViewModelConverter sequenceViewModelConverter;
    private final SequenceSearchViewModelConverter sequenceSearchViewModelConverter;
    private final SequenceChannelViewModelConverter sequenceChannelViewModelConverter;
    private final StageViewModelConverter stageViewModelConverter;
    private final StagePropViewModelConverter stagePropViewModelConverter;

    @Inject
    SequenceRestServiceHandler(SequencePersistenceService sequencePersistenceService,
                               SongPersistenceService songPersistenceService,
                               StagePersistenceService stagePersistenceService,
                               SequenceViewModelConverter sequenceViewModelConverter,
                               SequenceSearchViewModelConverter sequenceSearchViewModelConverter,
                               SequenceChannelViewModelConverter sequenceChannelViewModelConverter,
                               StageViewModelConverter stageViewModelConverter,
                               StagePropViewModelConverter stagePropViewModelConverter) {
        this.sequencePersistenceService = sequencePersistenceService;
        this.songPersistenceService = songPersistenceService;
        this.stagePersistenceService = stagePersistenceService;
        this.sequenceViewModelConverter = sequenceViewModelConverter;
        this.sequenceSearchViewModelConverter = sequenceSearchViewModelConverter;
        this.sequenceChannelViewModelConverter = sequenceChannelViewModelConverter;
        this.stageViewModelConverter = stageViewModelConverter;
        this.stagePropViewModelConverter = stagePropViewModelConverter;
    }

    @Transactional
    Response createSequence(SequenceViewModel sequenceViewModel) {
        sequenceViewModel.setId(null);
        sequenceViewModel.setStatus(SequenceStatus.NEW);

        Sequence sequence = sequenceViewModelConverter.toModel(sequenceViewModel);
        sequence = sequencePersistenceService.createSequence(sequence);
        return respondOk(new IdResponse(sequence.getId()));
    }

    Response getAllSequences() {
        List<Sequence> sequences = sequencePersistenceService.getAllSequences();
        List<SequenceSearchViewModel> viewModels = sequenceSearchViewModelConverter.toViewModels(sequences);
        return respondOk(viewModels);
    }

    Response getSequence(int sequenceId) {
        Optional<Sequence> sequenceOptional = sequencePersistenceService.getSequenceById(sequenceId);

        if (sequenceOptional.isPresent()) {
            Sequence sequence = sequenceOptional.get();
            SequenceViewModel viewModel = sequenceViewModelConverter.toViewModel(sequence);

            List<SequenceChannelViewModel> channels = sequencePersistenceService
                    .getSequenceChannelsBySequenceId(sequenceId)
                    .stream()
                    .map(sequenceChannelViewModelConverter::toViewModel)
                    .collect(toList());
            viewModel.setChannels(channels);

            return respondOk(viewModel);
        }

        return respond(Response.Status.NOT_FOUND, "Sequence not found.");
    }

    Response getSequenceStage(int sequenceId) {
        Optional<Stage> stageOptional = sequencePersistenceService.getStageBySequenceId(sequenceId);

        if (stageOptional.isPresent()) {
            Stage stage = stageOptional.get();

            StageViewModel viewModel = stageViewModelConverter.toViewModel(stage);
            List<StagePropViewModel> stageProps = stagePersistenceService
                    .getStagePropsByStageId(stage.getId())
                    .stream()
                    .map(stagePropViewModelConverter::toViewModel)
                    .collect(toList());
            viewModel.setStageProps(stageProps);
            return respondOk(viewModel);
        } else {
            return respond(Response.Status.NOT_FOUND, "Stage not found.");
        }
    }

    Response getSequenceSongAudio(int sequenceId) {
        Optional<SongAudio> songAudioOptional = sequencePersistenceService.getSongAudioBySequenceId(sequenceId);
        if (songAudioOptional.isPresent()) {
            return respondMedia(songAudioOptional.get().getAudioData(), MP3_MIME_TYPE);
        } else {
            return respond(Response.Status.NOT_FOUND, "Song audio not found.");
        }
    }

    @Transactional
    Response updateSequence(int id, SequenceViewModel sequenceViewModel) {
        sequenceViewModel.setId(id); // Prevent client-side ID tampering.

        Sequence sequence = sequenceViewModelConverter.toModel(sequenceViewModel);
        List<SequenceChannel> sequenceChannels = sequenceViewModel.getChannels()
                .stream()
                .map(sequenceChannelViewModelConverter::toModel)
                .collect(toList());

        if (sequence.getStatus() == SequenceStatus.PUBLISHED) {
            publishSequence(sequence, sequenceChannels);
        } else {
            saveDraftSequence(sequence, sequenceChannels);
        }

        return respondOk();
    }

    @Transactional
    Response deleteSequence(int id) {
        sequencePersistenceService.deleteSequence(id);
        return respondOk();
    }

    private void saveDraftSequence(Sequence sequence, List<SequenceChannel> sequenceChannels) {
        sequencePersistenceService.saveSequence(sequence, sequenceChannels);
    }

    private void publishSequence(Sequence sequence, List<SequenceChannel> sequenceChannels) {
        Song song = songPersistenceService.getSongBySequenceId(sequence.getId())
                .orElseThrow(() -> new EntityNotFoundException("Song not found."));

        List<StageProp> stageProps = stagePersistenceService.getStagePropsByStageId(sequence.getStageId());
        int endFrame = SequenceUtils.getFrameCount(song, sequence) - 1;

        RenderedStagePropDataMap renderedStageProps = new Renderer(sequence, sequenceChannels, stageProps, 0, endFrame).render();
        sequencePersistenceService.publishSequence(sequence, sequenceChannels, renderedStageProps);
    }
}

File: rest/src/main/kotlin/io/sparkled/rest/service/sequence/SequenceRestServiceHandler.kt
package io.sparkled.rest.service.sequence

import io.sparkled.model.animation.SequenceChannelEffects
import io.sparkled.model.entity.Sequence
import io.sparkled.model.entity.SequenceChannel
import io.sparkled.model.entity.SequenceStatus
import io.sparkled.model.render.RenderResult
import io.sparkled.model.util.SequenceUtils
import io.sparkled.model.validator.exception.EntityNotFoundException
import io.sparkled.persistence.sequence.SequencePersistenceService
import io.sparkled.persistence.song.SongPersistenceService
import io.sparkled.persistence.stage.StagePersistenceService
import io.sparkled.persistence.transaction.Transaction
import io.sparkled.renderer.Renderer
import io.sparkled.rest.response.IdResponse
import io.sparkled.rest.service.RestServiceHandler
import io.sparkled.viewmodel.sequence.SequenceViewModel
import io.sparkled.viewmodel.sequence.SequenceViewModelConverter
import io.sparkled.viewmodel.sequence.channel.SequenceChannelViewModelConverter
import io.sparkled.viewmodel.sequence.search.SequenceSearchViewModelConverter
import io.sparkled.viewmodel.stage.StageViewModelConverter
import io.sparkled.viewmodel.stage.prop.StagePropViewModelConverter
import org.slf4j.LoggerFactory
import java.util.UUID
import javax.inject.Inject
import javax.ws.rs.core.Response
import kotlin.math.min

open class SequenceRestServiceHandler
@Inject constructor(
    private val transaction: Transaction,
    private val sequencePersistenceService: SequencePersistenceService,
    private val songPersistenceService: SongPersistenceService,
    private val stagePersistenceService: StagePersistenceService,
    private val sequenceViewModelConverter: SequenceViewModelConverter,
    private val sequenceSearchViewModelConverter: SequenceSearchViewModelConverter,
    private val sequenceChannelViewModelConverter: SequenceChannelViewModelConverter,
    private val stageViewModelConverter: StageViewModelConverter,
    private val stagePropViewModelConverter: StagePropViewModelConverter
) : RestServiceHandler() {

    fun createSequence(sequenceViewModel: SequenceViewModel): Response {
        return transaction.of {
            sequenceViewModel.setId(null)
            sequenceViewModel.setStatus(SequenceStatus.NEW)

            val sequence = sequenceViewModelConverter.toModel(sequenceViewModel)
            val sequenceChannels = createSequenceChannels(sequence)
            val savedSequence = sequencePersistenceService.saveSequence(sequence, sequenceChannels)

            return@of respondOk(IdResponse(savedSequence.getId()!!))
        }
    }

    private fun createSequenceChannels(sequence: Sequence): List<SequenceChannel> {
        val stageProps = stagePersistenceService.getStagePropsByStageId(sequence.getStageId()!!)
        return stageProps.mapIndexed { i, it ->
            SequenceChannel()
                .setUuid(UUID.randomUUID())
                .setName(it.getName())
                .setStagePropUuid(it.getUuid())
                .setDisplayOrder(i)
                .setChannelJson(SequenceChannelEffects.EMPTY_JSON)
        }
    }

    fun getAllSequences(): Response {
        val sequences = sequencePersistenceService.getAllSequences()
        val viewModels = sequenceSearchViewModelConverter.toViewModels(sequences)
        return respondOk(viewModels)
    }

    fun getSequence(sequenceId: Int): Response {
        val sequence = sequencePersistenceService.getSequenceById(sequenceId)

        if (sequence != null) {
            val viewModel = sequenceViewModelConverter.toViewModel(sequence)

            val channels = sequencePersistenceService
                .getSequenceChannelsBySequenceId(sequenceId)
                .asSequence()
                .map(sequenceChannelViewModelConverter::toViewModel)
                .toList()
            viewModel.setChannels(channels)

            return respondOk(viewModel)
        }

        return respond(Response.Status.NOT_FOUND, "Sequence not found.")
    }

    fun getSequenceStage(sequenceId: Int): Response {
        val stage = sequencePersistenceService.getStageBySequenceId(sequenceId)

        return if (stage != null) {
            val viewModel = stageViewModelConverter.toViewModel(stage)
            val stageProps = stagePersistenceService
                .getStagePropsByStageId(stage.getId()!!)
                .asSequence()
                .map(stagePropViewModelConverter::toViewModel)
                .toList()

            viewModel.setStageProps(stageProps)
            respondOk(viewModel)
        } else {
            respond(Response.Status.NOT_FOUND, "Stage not found.")
        }
    }

    fun getSequenceSongAudio(sequenceId: Int): Response {
        val songAudio = sequencePersistenceService.getSongAudioBySequenceId(sequenceId)
        return if (songAudio != null) {
            respondMedia(songAudio.getAudioData()!!, MP3_MIME_TYPE)
        } else {
            respond(Response.Status.NOT_FOUND, "Song audio not found.")
        }
    }

    fun previewSequence(
        sequenceID: Int,
        startFrame: Int,
        frameCount: Int,
        sequenceViewModel: SequenceViewModel
    ): Response {
        logger.info("Previewing sequence #$sequenceID.")
        val sequence = sequenceViewModelConverter.toModel(sequenceViewModel)

        val sequenceChannels = sequenceViewModel.getChannels()
            .asSequence()
            .map(sequenceChannelViewModelConverter::toModel)
            .toList()

        return respondOk(
            renderSequence(sequence, sequenceChannels, startFrame, startFrame + frameCount - 1)
        )
    }

    fun updateSequence(id: Int, sequenceViewModel: SequenceViewModel): Response {
        return transaction.of {
            sequenceViewModel.setId(id) // Prevent client-side ID tampering.

            val sequence = sequenceViewModelConverter.toModel(sequenceViewModel)
            val sequenceChannels = sequenceViewModel.getChannels()
                .asSequence()
                .map(sequenceChannelViewModelConverter::toModel)
                .toList()

            if (sequence.getStatus() === SequenceStatus.PUBLISHED) {
                publishSequence(sequence, sequenceChannels)
            } else {
                saveDraftSequence(sequence, sequenceChannels)
            }

            return@of respondOk()
        }
    }

    fun deleteSequence(id: Int): Response {
        return transaction.of {
            sequencePersistenceService.deleteSequence(id)
            return@of respondOk()
        }
    }

    private fun saveDraftSequence(sequence: Sequence, sequenceChannels: List<SequenceChannel>) {
        sequencePersistenceService.saveSequence(sequence, sequenceChannels)
    }

    private fun publishSequence(sequence: Sequence, sequenceChannels: List<SequenceChannel>) {
        val renderResult = renderSequence(sequence, sequenceChannels)
        sequencePersistenceService.publishSequence(sequence, sequenceChannels, renderResult.stageProps)
    }

    private fun renderSequence(
        sequence: Sequence,
        sequenceChannels: List<SequenceChannel>,
        startFrame: Int = 0,
        endFrame: Int = Int.MAX_VALUE
    ): RenderResult {
        val song = songPersistenceService.getSongBySequenceId(sequence.getId()!!)
            ?: throw EntityNotFoundException("Song not found.")

        val stageProps = stagePersistenceService.getStagePropsByStageId(sequence.getStageId()!!)
        val endFrameBounded = min(endFrame, SequenceUtils.getFrameCount(song, sequence) - 1)

        return Renderer(sequence, sequenceChannels, stageProps, startFrame, endFrameBounded).render()
    }

    companion object {
        private const val MP3_MIME_TYPE = "audio/mpeg"
        private val logger = LoggerFactory.getLogger(SequenceRestServiceHandler::class.java)
    }
}